.github/workflows/src/verify-run-status.js

```diff
@@ -0,0 +1,161 @@
+import { PER_PAGE_MAX } from "./github.js";
+import { extractInputs } from "./context.js";
+
+/**
+ *
+ * @param {import('github-script').AsyncFunctionArguments} AsyncFunctionArguments
+ * @param {string} checkRunName
+ * @param {string} workflowName
+ */
+export async function verifyRunStatus(
+  { github, context, core },
+  checkRunName,
+  workflowName,
+) {
+  // Exit early when context is a check_run event and the check run does not
+  // match the checkRunName.
+  if (context.eventName == "check_run") {
+    const contextRunName = context.payload.check_run.name;
+    if (contextRunName !== checkRunName) {
+      core.info(
+        `Check run name (${contextRunName}) does not match input: ${checkRunName}`,
+      );
+      return;
+    }
+  }
+
+  const { head_sha } = await extractInputs(github, context, core);
+
+  const checkRun =
+    context.eventName == "check_run"
+      ? context.payload.check_run
+      : await getCheckRunStatus(github, context, core, checkRunName, head_sha);
+
+  if (!checkRun) {
+    core.info(`No completed check run with name: ${checkRunName}`);
+    return;
+  }
+
+  core.info(
+    `Check run name: ${checkRun.name}, conclusion: ${checkRun.conclusion}, URL: ${checkRun.html_url}`,
+  );
+  core.debug(`Check run: ${JSON.stringify(checkRun)}`);
+
+  const workflowRun =
+    context.eventName == "workflow_run"
+      ? context.payload.workflow_run
+      : await getWorkflowRun(github, context, core, workflowName, head_sha);
+
+  if (!workflowRun) {
+    core.info(`No completed workflow run with name: ${workflowName}`);
+    return;
+  }
+
+  core.info(
+    `Workflow run name: ${workflowRun.name}, conclusion: ${workflowRun.conclusion}, URL: ${workflowRun.html_url}`,
+  );
+  core.debug(`Workflow run: ${JSON.stringify(workflowRun)}`);
+
+  if (checkRun.conclusion !== workflowRun.conclusion) {
+    core.setFailed(
+      `Check run conclusion (${checkRun.conclusion}) does not match workflow run conclusion (${workflowRun.conclusion})`,
+    );
+    return;
+  }
+
+  core.info("Checks match");
+}
+
+/**
+ *
+ * @param {import('github-script').AsyncFunctionArguments['github']} github
+ * @param {import('github-script').AsyncFunctionArguments['context']} context
+ * @param {import('github-script').AsyncFunctionArguments['core']} core
+ * @param {string} checkRunName
+ * @param {string} head_sha
+ */
+export async function getCheckRunStatus(
```

Comment:
I think `getCheckRunStatus()` and `getWorkflowRun()` should be moved to shared helpers, since other workflows will need them.  We will need to think about the best way to factor the APIs, including extractInputs().

Reply:
Discussion results: 
* There is some useful abstraction for getting check runs and workflow info beyond wrapping `github.paginate` and some basic validation/sorting
* `extractInputs` should return lazy or getters/setters that do lazy loading/caching of results so only desired inputs require 

https://github.com/Azure/azure-rest-api-specs/issues/33772